contracts/TokenListManager.sol

```diff
@@ -6,176 +6,80 @@ import "@openzeppelin/contracts/utils/structs/EnumerableSet.sol";
-    mapping(address => bool) private whitelistedTokens;
-    // Use mapping to check if token in blacklist
-    mapping(address => bool) private blacklistedTokens;
-
-    // Set permissions
-    modifier onlyOwner() {
-        // Match address
-        /** 0xC0Ec75c25EC201885a791Fd8d39Bf8CE96e1c566
-         * this address is my own test address
-         * owner can use thier own address to replace. */
-        require(msg.sender == 0xC0Ec75c25EC201885a791Fd8d39Bf8CE96e1c566, "Not authorized");
-        _;
+import "@openzeppelin/contracts/access/AccessControl.sol";
+
+contract TokenListManager is ITokenListManager, AccessControl {
+
+    // Utilize EnumerableSet for efficient set operations.
+    using EnumerableSet for EnumerableSet.AddressSet;
+
+    // Define roles for administrative control and user access.
+    bytes32 public constant ADMIN_ROLE = keccak256("ADMIN_ROLE");
+    bytes32 public constant USER_ROLE = keccak256("USER_ROLE");
+
+    // Mappings for whitelisted and blacklisted tokens, categorized by token type.
+    mapping(TokenType => EnumerableSet.AddressSet) private whitelistedTokens;
+    mapping(TokenType => EnumerableSet.AddressSet) private blacklistedTokens;
+
+    // Contract constructor setting up initial admin role.
+    constructor() {
+        _grantRole(DEFAULT_ADMIN_ROLE, msg.sender);
+        _setRoleAdmin(USER_ROLE, ADMIN_ROLE);
     }
 
-    // Implement the addTokens method and add tokens to whitelist
-    function addTokens(address[] calldata tokens) external override onlyOwner {
-        for (uint i = 0; i < tokens.length; i++) {
-            // Check if the token is not already whitelisted
-            if (!whitelistedTokens[tokens[i]]) {
-                // If the token type is ERC-20
-                if (IERC165(tokens[i]).supportsInterface(type(IERC20).interfaceId)) {
-                    // Storage to whitelistErc20Tokens
-                    whitelistErc20Tokens.push(tokens[i]);
-                // If the token type is IERC-721
-                } else if (IERC165(tokens[i]).supportsInterface(type(IERC721).interfaceId)) {
-                    // Storage to whitelistErc721Tokens
-                    whitelistErc721Tokens.push(tokens[i]);
-                // If the token type is IERC-1155
-                } else if (IERC165(tokens[i]).supportsInterface(type(IERC1155).interfaceId)) {
-                    // Storage to whitelistErc1155Tokens
-                    whitelistErc1155Tokens.push(tokens[i]);
-                } else {
-                    revert("Unsupported token type");
-                }
-                // Mark the token address as whitelisted in the mapping
-                whitelistedTokens[tokens[i]] = true;
-                // Add the token address to the array of whitelisted tokens
-                whitelistedTokensArray.push(tokens[i]);
-            }
-        }
+    // Modifier to restrict function access to users with the admin role.
+    modifier onlyAdmin() {
```

Comment:
Could reuse the `hasRole` modifier in `AccessControl`.

Reply:
Deleted